### DISCUSSION TITLE
fetch full articles behind paywall

### DISCUSSION THREAD

**Discussion by ilf** [spacecowboy/Feeder#220] [Ideas]
Unfortunately, more and more news websites hide more and more content behind paywalls.

Many still offer feeds, listing all articles with titles and abstracts, for example:

- https://www.spiegel.de/schlagzeilen/tops/index.rss
- https://rss.nytimes.com/services/xml/rss/nyt/HomePage.xml

Fetching them as "full articles" will work for some articles without paywall, but will fail for other  articles behind paywall.

A Feeder user might have a login for those paywalls, though. But currently, using that requires leaving Feeder, opening a browser, logging in there, and opening the article URL in the browser instead of Feeder.

It would be awesome if we could integrate logins to paywalls into Feeder, to fetch the full articles even behind paywalls.

Maybe we could achieve this by just storing the username/password? Or an authentication cookie gained after logging in the particular site?

I don't have the complete solution. But this would be highly useful!

Thanks, and keep up the good work!

**Comment by spacecowboy** [comment]
Nothing to be done in Feeder. A paywall is there for a reason.

For sites where you can log in, I recommend configuring them to open by default with `custom tab`. This effectively uses your default browser, where you will have already logged in etc but you won't "leave" Feeder.

**Comment by ilf** [comment]
I'm not convinced opening articles in "custom tab" is the right solution.

I am already working around paywalls by opening an external browser. One reason I want to change that is to have a persitant login in Feeder, but not in my browser. I regularly delete "browsing data" like cookies in my external browser.

Besides not solving my problem, displaying the browser as a "custom tab" instead of "default browser" also creates new issues: First, this setting is not only for articles, but all links. Second, firefox in "custom tab" doesn't allow using its reader mode.

> Nothing to be done in Feeder. A paywall is there for a reason.

I agree with the second sentence, but not with the first. Feeder already choses to "enhance" what publishers and feeds offer, by providing the awesome feature "fetch full articles (by default)".

My suggestion would build on that: a "full articles" feature with cookies (or other suitable login information).

**Reply by spacecowboy** [reply]
Well feel free to hack on it and see if you could contribute something.

**Reply by spacecowboy** [reply]
just realize that what you are imagining is likely a pretty huge engineering effort.  you are essentially suggesting to build a custom browser.

**Comment by ilf** [comment]
One more argument for this feature: I don't need to be online for opening and reading the article, only for fetching it (which is done periodically in background).

I'm not sure it's a huge engineering effort.

The paywall on spiegel.de can be passed with a simple HTTP header `Cookie: accessInfo=TOKEN`. In cURL, that's `-H 'Cookie: accessInfo=TOKEN'`.

I'm pretty sure that would work for a lot of paywalls.

I'm not requesting an entire system to do an actual login on a target website, analyze how and where the authorization is stored, and then extract it and export it into Feeder.

But how about a simple feed option to specificy a cookie key and value?

**Reply by spacecowboy** [reply]
If you want it, build it and see if it works.

I have no interest in building it

**Reply by ilf** [reply]
Thanks for the offer. But I am not a developer, and I have never heard of Kotlin before I tried to look into the source code just now.

I assume the additional HTTP header would be set somewhere in https://github.com/spacecowboy/Feeder/blob/master/app/src/main/java/com/nononsenseapps/feeder/sync/FeederSync.kt?

**Reply by spacecowboy** [reply]
There's more than one way to do it. For a quick experiment, where the header is set for all requests, it can be done here:

https://github.com/spacecowboy/Feeder/blob/master/app/src/main/java/com/nononsenseapps/feeder/FeederApplication.kt#L115

by adding another `NetworkInterceptor`

and a `NetworkInterceptor` is pretty simple, this could be used as a template:
https://github.com/spacecowboy/Feeder/blob/master/app/src/main/java/com/nononsenseapps/feeder/model/UserAgentInterceptor.kt

But as for a "real" solution, it would probably have to be conditional? Is there a downside you think in sending this header to all sites?

**Reply by spacecowboy** [reply]
Also, doesn't this header only say "Please look in the cookie for this info"? And what happens when there is no cookie?

**Reply by ilf** [reply]
Yes, it would be wise to send an access token cookie only to the site it's valid for. It's not only useless on other sites, also they could copy and abuse it. I think setting it as an option per feed would be ideal.

The HTTP header `Cookie: key=value` *is* the (content of the) cookie. That's the beauty of it, it's really simple. :)

**Reply by LMSSonos** [reply]
@spacecowboy I tested it quickly the URL mentioned here https://github.com/spacecowboy/Feeder/discussions/345 (same topic but cookie consent instead of paywall) - and worked as expected.

But I'm not really sure what the best way is to implement it, especially I think it's not really easy for the user to take the cookie and paste it. So I was thinking that I could launch a webview and then grab the cookie from there (https://stackoverflow.com/questions/11100086/android-extracting-cookies-after-login-in-webview)

What do you think if I add a option in the settings like "Get auth/consent cookie from site", then you click the button and then simply would load the webview, once the user closes it, I save the cookie with the feed? Or do you have any better idea or do not like the idea at all ;-)

**Reply by spacecowboy** [reply]
Would it be possible to hardcode these headers in the app?

The naive way would be to hardcode known headers and just send them with every request.

If that breaks, maybe have a list or URLs and known required HTTP headers.

Or are the cookies required to be valid session values or similar?

**Reply by ilf** [reply]
I don't know about #345 or cookie banners.

This issue is about paywalls. These can't (easily and legally) be circumvented by a hardcoded header shared across all users.

These headers would be restricted to accounts. See my previous posts:

> Maybe we could achieve this by just storing the username/password? Or an authentication cookie gained after logging in the particular site?

> The paywall on spiegel.de can be passed with a simple HTTP header Cookie: accessInfo=TOKEN. In cURL, that's -H 'Cookie: accessInfo=TOKEN'.

**Reply by LMSSonos** [reply]
@spacecowboy - finally found the time and implemented just some code to see if the logic would work. so basically I did

1) Create a new Activity with `WebView` in it
2) In the Feed I added a new MenuIem which opens `Webview`
3) Once opened, you can login, accept the cookies, etc
4) When the WebView is closed, I catch the cookies
5) In the `FullTextParser.kt` I changed the `okHttpClient` to include the cookies

Things I did not handle: Save the cookie with the feed settings. We should also consider that these cookies may expire over time. 

Question, do you prefer to have a specific implementation, like instead of the MenuItem move it to the Feed Settings? As I never made anything in Kotlin/Android, do you want me to share my dummy implementation with a lot of hacks or do you want to get a PR from me - just be warned, that probably some stuff is not implemented as it should be, but I would be happy to change the findings.

**Reply by spacecowboy** [reply]
> Question, do you prefer to have a specific implementation, like instead of the MenuItem move it to the Feed Settings? As I never made anything in Kotlin/Android, do you want me to share my dummy implementation with a lot of hacks or do you want to get a PR from me - just be warned, that probably some stuff is not implemented as it should be, but I would be happy to change the findings.

Feel free to share what you did. Seeing what you did will definitely influence how I think it should be done "for real" if implemented.

**Reply by LMSSonos** [reply]
@spacecowboy you can find it here: https://github.com/LMSSonos/feeder-cookie/commit/dcc01e75da522abe0c75b2d397a81a7be1b4d7e0 

1) It starts in ArticleScreen.kt (Load Cookies DropdownMenuItem). There I have hardcoded the URL. IMHO this should be moved to the feed settings and then launch the WebViewActivity
2) Once the WebViewActivity is closed, in the same file i store `cookies = WebViewActivity.extractCookies(result.data)` the cookies. I think, this should be stored in the DB
3) In FullTextParser.kt you see, that I use the cookie and send it with the webrequest

In this demo, you need everytime to go into the ArticleDetail and then open the WebView (Cookie) and then Load the full article. 

As you can see, there still remains a lot of work to implement this nicely ;-)
